### PR TITLE
feat(apps): Make `setupPlugin` errors `RetryError`-only and send "fatal app error" email

### LINKED ISSUE #6850: Send a PagerDuty alert when a plugin gets autodisabled
Currently when a plugin errors, it gets autodisabled. This is however completely silent and easy to miss. The result can be degraded data quality or a hole in data in case of export plugins – something to be fixed ASAP.

A PagerDuty integration could come in useful here, allowing an alert to be sent in case of a plugin getting disabled. PagerDuty is tailored for this kind of thing, so this would potentially be better than e.g. an email – that could be easily missed in a sea of emails.

This problem was brought up by a current user of our Snowflake Export Plugin.

### PR DESCRIPTION
**Pull request by Twixes**
## Changes

Resolves #6850 by sending an email whenever a ~~plugin~~ app is disabled by the system due to a VM initialization error (usually in `setupPlugin`). Also makes it so that only `RetryError` result in `setupPlugin` being retried. Both of these are part of the https://github.com/PostHog/posthog/issues/8829#issuecomment-1147363646 action plan.

err

We only send emails from Python and Celery is the best way to queue emails, but at the same time I really wanted to avoid reverting #9936, so instead I read up on the Celery task message protocol + did some reverse engineering, and ended up adding a small method to `DB` called `celeryApplyAsync()`. It works pretty much like `Task.apply_async()` in Python, just with no bells and whistles.

## How did you test this code?

~~Working on some unit tests.~~ Added tests to `vm.lazy.test.ts`.

### CODE REVIEW

**Review by tiina303 — APPROVED** [review]

File: posthog/templates/email/fatal_plugin_error.html [hunk 1]
```diff
@@ -0,0 +1,26 @@
+{% extends "email/base_transactional.html" %} {% load posthog_assets %} {% load posthog_filters %}
+{% block heading %}Fatal app error{% endblock %}
+{% block section %}
+<!-- prettier-ignore -->
+<p>
+    {% if is_system_error %}
+        A system error has prevented {{ plugin }}
+    {% else %}
+        An unhandled error in the <code>setupPlugin</code> function of {{ plugin }} has prevented it
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/templates/email/fatal_plugin_error.html:9]: I think there needs to be a broader strategy, as official apps still call themselves "plugins", we recommend using a `plugin-scaffold` package, the events are processed by the "plugin server", and overall the word is still in a lot of places. In this PR I mostly handled this team/projects style – UI and messaging uses "app", code uses "plugin".

File: plugin-server/src/worker/vm/lazy.ts [hunk 2]
```diff
@@ -192,39 +191,46 @@ export class LazyPluginVM {
...
+                    PluginLogEntryType.Error
+                )
+                this.initRetryTimeout = setTimeout(async () => {
+                    await this._setupPlugin(vm)
+                }, nextRetryMs)
+            } else {
+                this.inErroredState = true
+                const errorInfo =
+                    error instanceof RetryError
+                        ? `RetryError (attempt ${VM_INIT_MAX_RETRIES}/${VM_INIT_MAX_RETRIES})`
```
[the remark this answers is not in the document]
Comment (Twixes) [reply plugin-server/src/worker/vm/lazy.ts:225]: Good idea, updated `RetryError` in https://github.com/PostHog/plugin-scaffold to facilitate this.

File: plugin-server/src/utils/db/db.ts [hunk 3]
```diff
@@ -432,6 +433,37 @@ export class DB {
...
     }
 
+    /** Calls Celery task. Works similarly to Task.apply_async in Python. */
+    async celeryApplyAsync(taskName: string, args: any[] = [], kwargs: Record<string, any> = {}): Promise<void> {
+        const taskId = new UUIDT().toString()
+        const deliveryTag = new UUIDT().toString()
+        const body = [args, kwargs, { callbacks: null, errbacks: null, chain: null, chord: null }]
+        /** A base64-encoded JSON representation of the body tuple. */
+        const bodySerialized = Buffer.from(JSON.stringify(body)).toString('base64')
+        await this.redisLPush(CELERY_DEFAULT_QUEUE, {
```
[the remark this answers is not in the document]
Comment (Twixes) [reply plugin-server/src/utils/db/db.ts:443]: Not any higher than with the heap of Celery client code we had before, as the task message protocol is constant.

File: posthog/templates/email/fatal_plugin_error.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (tiina303) [reply posthog/templates/email/fatal_plugin_error.html:9]: Do we have an issue to track this (like we had for FOSS)?

File: posthog/tasks/email.py [hunk 4]
```diff
@@ -63,6 +62,35 @@ def send_member_join(invitee_uuid: str, organization_id: str) -> None:
...
+    )
+    # Don't send this email to the new member themselves
+    memberships_to_email = list(
+        filter(
+            team.get_effective_membership_level_for_parent_membership,
+            OrganizationMembership.objects.select_related("user", "organization").filter(
+                organization_id=team.organization_id
+            ),
+        )
+    )
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/tasks/email.py]: Rewrote this to use list comprehension, which should be more obvious

File: posthog/tasks/email.py [hunk 5]
```diff
@@ -77,14 +77,13 @@ def send_fatal_plugin_error(
...
-            OrganizationMembership.objects.select_related("user", "organization").filter(
-                organization_id=team.organization_id
-            ),
+    memberships_to_email = [
+        membership
+        for membership in OrganizationMembership.objects.select_related("user", "organization").filter(
+            organization_id=team.organization_id
         )
-    )
+        if team.get_effective_membership_level_for_parent_membership(membership) is not None
```
Comment (tiina303) [inline posthog/tasks/email.py:85]: maybe a comment to this function or there that we return `None` for which case

File: plugin-server/src/worker/vm/lazy.ts [hunk 6]
```diff
@@ -192,39 +191,42 @@ export class LazyPluginVM {
...
-            )
-            await this.createLogEntry(error.message, PluginLogEntryType.Error)
             this.clearRetryTimeoutIfExists()
-            this.lastError = error
-            this.initRetryTimeout = setTimeout(async () => {
-                await this._setupPlugin(vm)
-            }, nextRetryMs)
+            if (error instanceof RetryError) {
+                error._attempt = this.totalInitAttemptsCounter
+                error._maxAttempts = VM_INIT_MAX_RETRIES
```
Comment (tiina303) [inline plugin-server/src/worker/vm/lazy.ts:206]: where are these used?

**Review by tiina303 — APPROVED** [review]

File: posthog/templates/email/fatal_plugin_error.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/templates/email/fatal_plugin_error.html:9]: I don't think so, though I haven't been involved in the rebranding

**Pull request merged**